content/purgatorio/launch-and-mantain-the-system/automation-and-reproducibility.md

```diff
@@ -1,5 +1,242 @@
+    - [Challenges in irreproducible research](https://www.nature.com/collections/prbfkwmwvz)
+
+- **Model training and serving**
+
+    You can create scripts to _automatically_ train ML models at any given time, perhaps using new data you have collected, cleaned, and pre-processed. 
+
+    You can automatically check that the performance of the models reflects expectations, and train them again in case they degrade.
+    
+    See:
+    - [Why Machine Learning Models Degrade In Production](https://towardsdatascience.com/why-machine-learning-models-degrade-in-production-d0f2108e9214)
+    - [The Ultimate Guide to Model Retraining](https://mlinproduction.com/model-retraining/)
+
+    Once you've retrained the model, you can _automatically_ deploy it in production. Many Cloud Providers (AWS, Azure, GCP) offer you the possibility to do this in a very simple way.
+    
+#### The Automation Golden Rule
+
+As you can see, there's a big room for improvement in your Data Science work if you use automation.
+
+The [Rule of Three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) in software development says: 
+
+> It states that two instances of similar code don't require refactoring, but when a similar code is used three times, it should be extracted into a new procedure.
+
+_Likewise_, you can think of automation!
+
+**If you should take a step more than twice, it's probably worth taking a moment to automate it!**
+
+## AutoML
+
+Until now we have seen how automation can fit into the various steps of the Data Science process, speed them up, and how it is vital to obtain reproducible results. 
+
+But there is another aspect of how automation can help us, and it is called [AutoML](https://en.wikipedia.org/wiki/Automated_machine_learning). 
+
+Automated machine learning (AutoML) is the process of automating the process of applying machine learning to real-world problems.
+
+**AutoML covers the complete pipeline from the raw dataset to the deployable machine learning model.**
+
+From Wikipedia:
+
+> In a typical Machine Learning application, practitioners have a set of input data points to train on. The raw data may not be in a form that all algorithms can be applied to it. To make the data amenable for machine learning, an expert may have to apply appropriate data pre-processing, feature engineering, feature extraction, and feature selection methods. After these steps, practitioners must then perform algorithm selection and hyperparameter optimization to maximize the predictive performance of their model. 
+
+> All of these steps induce challenges, accumulating a significant hurdle to get started with machine learning.
+
+> **AutoML** dramatically simplifies these steps for non-experts. Automation is not perfect because AutoML tools still have hyperparameters, and setting them may require expertise.
+
+Google is a pioneer in AutoML, and engineers there wrote a log of blog posts about this approach. Take a read: 
```

Comment:
In my view, H2O's Driverless AI beats many tools out there and there are many advantages of use this over other AutoML tools, but this space is very competitive, you can see the list of tools under these two resources:
- [Things to know: AutoML](https://github.com/neomatrix369/awesome-ai-ml-dl/blob/master/things-to-know.md#automated-machine-learning-tools-or-partial-automl-tools)
- [Cloud, Infra, DevOps](https://github.com/neomatrix369/awesome-ai-ml-dl/blob/master/cloud-devops-infra/README.md#cloud-devops-infra)

Reply:
You will see some of those names have recently popped up and they are doing a great job, I recommend having a look at their tools.